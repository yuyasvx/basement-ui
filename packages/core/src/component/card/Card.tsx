import clsx from 'clsx';
import { forwardRef, type HTMLAttributes, type PropsWithChildren, useMemo } from 'react';
import { useVariant, type VariantAcceptable } from '../../hook/variant/VariantHook';
import type { Case } from '../../lib/Case';
import { ComponentToken } from '../ComponentToken';
import { calcCardShadow } from './CardShadowCalculator';
import { CardStyleVariable } from './CardStyleVariable';

export const CardVariant = {
  SHADOW: 'shadow',
  BORDER: 'border',
} as const;

export type CardStyleProps = {
  variantOption?: {
    shadowWidth?: number;
    shadowStrength?: number;
    borderWidth?: number;
    borderColor?: string;
  };
  baseColor?: string;
  radius?: number;
  backgroundAlpha?: number;
  backdropBlur?: number;
} & VariantAcceptable<typeof CardVariant>;

export type CardProps = PropsWithChildren<CardStyleProps & HTMLAttributes<HTMLDivElement>>;

export const Card = forwardRef<HTMLDivElement, CardProps>((props, ref) => {
  const { className, style } = props;
  const { newProps, restProps } = useCardStyle(props);
  newProps.className = useMemo(() => clsx(newProps.className, className), [className, newProps.className]);
  newProps.style = { ...newProps.style, ...style };

  return <div {...restProps} {...newProps} ref={ref} />;
});

export function useCardStyle<P extends CardStyleProps>(props: P) {
  const { variantOption, baseColor, radius, backgroundAlpha, backdropBlur, ...restProps1 } = props;

  const { variantClassName, restProps: restProps2 } = useVariant(restProps1, CardVariant.SHADOW as Case<typeof CardVariant>);

  const shadowValue = useMemo(
    () => (variantOption?.shadowWidth != null ? calcCardShadow(variantOption.shadowWidth, variantOption.shadowStrength) : undefined),
    [variantOption?.shadowStrength, variantOption?.shadowWidth],
  );
  const alphaPercent = backgroundAlpha != null ? percent(backgroundAlpha) : undefined;
  const cls = useMemo(
    () => clsx(ComponentToken.CARD, variantClassName, { '-alpha-override': backgroundAlpha != null }, { '-blur': backdropBlur != null }),
    [backdropBlur, backgroundAlpha, variantClassName],
  );
  const styleOverrideEnabled = useMemo(() => isStyleOverrideEnabled(props), [props]);

  return {
    newProps: {
      className: cls,
      style: styleOverrideEnabled
        ? {
            ...shadowValue?.cssStyle,
            [`--${CardStyleVariable.BASE_COLOR}`]: baseColor,
            [`--${CardStyleVariable.RADIUS}`]: radius != null ? `${radius}px` : undefined,
            [`--${CardStyleVariable.BASE_ALPHA}`]: alphaPercent != null ? `${alphaPercent}%` : undefined,
            [`--${CardStyleVariable.BACKDROP_BLUR}`]: backdropBlur != null ? `${backdropBlur}px` : undefined,
            [`--${CardStyleVariable.BORDER_WIDTH}`]: variantOption?.borderWidth != null ? `${variantOption?.borderWidth}px` : undefined,
            [`--${CardStyleVariable.BORDER_COLOR}`]: variantOption?.borderColor,
          }
        : undefined,
    },
    restProps: restProps2,
  };
}

Card.displayName = 'Card';

function percent(value: number) {
  const val = value > 1 ? 1 : value < 0 ? 0 : value;
  return val * 100;
}

function isStyleOverrideEnabled(props: CardStyleProps) {
  const { variantOption, baseColor, radius, backgroundAlpha, backdropBlur } = props;
  return baseColor != null || variantOption != null || radius != null || backgroundAlpha != null || backdropBlur != null;
}
