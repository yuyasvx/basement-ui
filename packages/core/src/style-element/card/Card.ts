import clsx from 'clsx';
import { useMemo } from 'react';
import { backgroundStyle, blurStyle, shadowStyle } from '../../domain/StyleClass';

export type BackgroundLevel = 0 | 1 | 2 | 3 | 4;
export type ShadowLevel = 0 | 1 | 2 | 3 | 4;
export type BlurLevel = 0 | 1;

const NAME = 'bm-a-card';
export const useCardStyle = (style?: CardStyle) => {
  const className = useMemo(
    () =>
      style == null
        ? ''
        : clsx([
            NAME,
            getBackgroundStyleClass(style.background),
            getShadowStyleClass(style.shadow),
            getBlurStyleClass(style.blur)
          ]),
    [style]
  );

  return {
    name: NAME,
    className,
    getBackgroundStyleClass,
    getShadowStyleClass,
    getBlurStyleClass
  };
};

export interface CardStyle {
  background: BackgroundLevel;
  blur: BlurLevel;
  shadow: ShadowLevel;
}

function getBackgroundStyleClass(level?: BackgroundLevel) {
  if (level == null) {
    return backgroundStyle[1];
  }
  if (level === 0) {
    return '';
  }
  return backgroundStyle[level - 1] ?? backgroundStyle[1];
}

function getShadowStyleClass(level?: ShadowLevel) {
  if (level == null || level === 0) {
    return '';
  }
  return shadowStyle[level - 1] ?? shadowStyle[1];
}

function getBlurStyleClass(level?: BlurLevel) {
  if (level == null || level === 0) {
    return '';
  }
  return blurStyle[0];
}
