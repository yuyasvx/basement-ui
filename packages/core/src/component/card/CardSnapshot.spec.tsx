import { render } from '@testing-library/react';
import { describe, expect, test } from 'vitest';
import { Card, CardVariant } from './Card';

describe('Card Snapshot', () => {
  test('Card Propsを何も指定しないとき、style属性のオーバーライドが発生しない', () => {
    const { container } = render(
      <>
        <Card>
          <div>content</div>
        </Card>
        <Card variantOption={{}}>
          <div>content</div>
        </Card>
      </>,
    );
    expect(container.childNodes).toMatchSnapshot();
  });

  test('Card Propsに何か指定すると、style属性のオーバーライドが発生する', () => {
    const { container } = render(
      <>
        <Card
          variantOption={{
            shadow: 10,
            borderColor: '#000',
          }}
        >
          <div>content</div>
        </Card>
        <Card radius={2}>
          <div>content</div>
        </Card>
        <Card backdropBlur={1}>
          <div>content</div>
        </Card>
        <Card backgroundAlpha={1}>
          <div>content</div>
        </Card>
        <Card baseColor="#000">
          <div>content</div>
        </Card>
      </>,
    );
    expect(container.childNodes).toMatchSnapshot();
  });

  test('Variantの指定が効いている', () => {
    const { container } = render(
      <>
        <Card variant={CardVariant.BORDER}>
          <div>content</div>
        </Card>
        <Card variant={CardVariant.SHADOW}>
          <div>content</div>
        </Card>
      </>,
    );
    expect(container.childNodes).toMatchSnapshot();
  });

  test('Card Propsは何も指定しないが、標準の属性（class, id, tabindexなど）を渡したとき、反映される', () => {
    const { container } = render(
      <Card className="test-class" id="card-id" tabIndex={1}>
        <div>content</div>
      </Card>,
    );
    expect(container.childNodes).toMatchSnapshot();
  });

  test('Card Props指定あり＆styleありの場合、統合される', () => {
    const { container } = render(
      <Card radius={6} variantOption={{ shadow: 5 }} style={{ width: '100px', borderRadius: '1px' }}>
        <div>content</div>
      </Card>,
    );
    expect(container.childNodes).toMatchSnapshot();
  });
});
