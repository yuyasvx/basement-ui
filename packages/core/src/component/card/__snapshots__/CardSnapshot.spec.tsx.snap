// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Card Snapshot > Card Propsに何か指定すると、style属性のオーバーライドが発生する 1`] = `
NodeList [
  <div
    class="bmui-card --shadow"
    style="--bmui-border-color: #000; --bmui-shadow-style: 0px 4px 10px color-mix(in srgb, #000000 16.000000000000004%, transparent), 0px 0px 2px color-mix(in srgb, #000000 25%, transparent);"
  >
    <div>
      content
    </div>
  </div>,
  <div
    class="bmui-card --shadow"
    style="--bmui-card-radius: 2px;"
  >
    <div>
      content
    </div>
  </div>,
  <div
    class="bmui-card --shadow -blur"
    style="--bmui-card-backdrop-blur: 1px;"
  >
    <div>
      content
    </div>
  </div>,
  <div
    class="bmui-card --shadow -alpha-override"
    style="--bmui-card-base-alpha: 100%;"
  >
    <div>
      content
    </div>
  </div>,
  <div
    class="bmui-card --shadow"
    style="--bmui-card-base-color: #000;"
  >
    <div>
      content
    </div>
  </div>,
]
`;

exports[`Card Snapshot > Card Propsは何も指定しないが、標準の属性（class, id, tabindexなど）を渡したとき、反映される 1`] = `
NodeList [
  <div
    class="bmui-card --shadow test-class"
    id="card-id"
    tabindex="1"
  >
    <div>
      content
    </div>
  </div>,
]
`;

exports[`Card Snapshot > Card Propsを何も指定しないとき、style属性のオーバーライドが発生しない 1`] = `
NodeList [
  <div
    class="bmui-card --shadow"
  >
    <div>
      content
    </div>
  </div>,
  <div
    class="bmui-card --shadow"
  >
    <div>
      content
    </div>
  </div>,
]
`;

exports[`Card Snapshot > Card Props指定あり＆styleありの場合、統合される 1`] = `
NodeList [
  <div
    class="bmui-card --shadow"
    style="--bmui-card-radius: 6px; --bmui-shadow-style: 0px 3px 5px color-mix(in srgb, #000000 15%, transparent), 0px 0px 2px color-mix(in srgb, #000000 27.499999999999996%, transparent); width: 100px; border-radius: 1px;"
  >
    <div>
      content
    </div>
  </div>,
]
`;

exports[`Card Snapshot > Variantの指定が効いている 1`] = `
NodeList [
  <div
    class="bmui-card --border"
  >
    <div>
      content
    </div>
  </div>,
  <div
    class="bmui-card --shadow"
  >
    <div>
      content
    </div>
  </div>,
]
`;
